Sensei Tailored Flow: Add 'Choose the purpose of your site' step

yscik opened the pull request:
Depends on https://github.com/Automattic/wp-calypso/pull/72473

#### Proposed Changes

* Add a new step after checkout to the Sensei flow, replicating the plugin onboarding screen
* Show dummy progress bar to pace out waiting for the atomic site
* Remove launchpad from steps

https://user-images.githubusercontent.com/176949/214653168-195835af-1c2d-46ba-9a13-2e45fc6b0bdf.mov

#### Testing Instructions

* Create a new site via http://calypso.localhost:3000/setup/sensei/
* After checkout, it should show the new screen
* To test just the screen without having to create a new site, use this link with an existing site URL and ID: `calypso.localhost:3000/setup/sensei/senseiPurpose?siteSlug= [URL] &siteId= [ID]`

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

## Screenshots for translation

    Order was completed successfully
    Preparing your new site

<img width="800" alt="image" src="https://user-images.githubusercontent.com/176949/217371611-f021b52b-6be0-4bf4-ba85-1d92ad7e695c.png">

    Choose the purpose of your site
    Select your goals for offering courses, and we will help you set everything up.
    Sell courses and generate income
    Provide certification
    Educate students
    Train employees

<img width="802" alt="image" src="https://user-images.githubusercontent.com/176949/217371887-df333486-17f2-4752-9993-ab03dd32cb8d.png">

yscik committed:
Add Purpose step

yscik committed:
Hide description until installing plugins is implemented

yscik committed:
Change redirect step after checkout

yscik committed:
Update navigation function in purpose step

yscik committed:
Turn off launchpad option

yscik committed:
Fix fallback pricing text

yscik committed:
Remove launchpad settings from launch step

yscik committed:
Follow-up refactors from rebase

gikaragia commented on client/landing/stepper/declarative-flow/internals/steps-repository/components/sensei-step-container/styles.scss:
```diff
@@ -1,3 +1,7 @@
+
+@import "@wordpress/base-styles/colors";
+@import "calypso/landing/stepper/declarative-flow/internals/sensei-variables";
```
I prefer the relative imports as it allows me to navigate directly to the source code. Is there a reason we should use "calypso/..."?

yscik replied:
Nope, just a mistake from moving files, updated in 478d920

gikaragia commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-purpose/index.tsx:
```diff
@@ -0,0 +1,162 @@
+import type { Step } from '../../types';
+import './style.scss';
+
+const wait = ( ms: number ) => new Promise( ( res ) => setTimeout( res, ms ) );
```
I don't think that defining a function for a single statement makes things clearer.

yscik replied:
This is a common little helper to make a promise-based timeout, I think it's clearer when using it just as `await wait( 500 );`

gikaragia replied:
Yeah, it's just that as I read the code, it takes me more time to navigate to the function to see what it does than it would take me to read a single statement.

In any case, I don't consider this a blocker.

gikaragia commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-purpose/index.tsx:
```diff
@@ -0,0 +1,162 @@
+		id: 'sell_courses',
+		label: __( 'Sell courses and generate income' ),
+		plugin: { slug: 'woocommerce', id: 'woocommerce/woocommerce' },
+		//description: __( 'WooCommerce will be installed for free.' ),
```
Commented out code

yscik replied:
Also cleaned this in 478d920. I only removed these description temporarily, since the next PR will be doing the installing :)

gikaragia reviewed (commented):
Left a couple of comments. Didn't get to testing yet.

yscik committed:
Fix code issues

gikaragia commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-purpose/index.tsx:
```diff
@@ -0,0 +1,158 @@
+		setFormState( ( formState ) => ( {
+			...formState,
+			selected: selected.includes( id )
+				? selected.filter( ( item ) => item !== id )
```
We could make this easier to read if we avoid generic variable names. Maybe 'newSelection' and 'oldSelection' or something similar?

gikaragia commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-purpose/style.scss:
```diff
@@ -0,0 +1,108 @@
+$breakpoint-mobile: 600px;
+$checked-color: #000;
+
+.sensei.sensei-purpose {
```
I think this is not used anywhere

yscik replied:
Do you mean the progress bar? That's for styling the progress at the top of the page.

gikaragia replied:
Ah my bad. Couldn't find the `sensei-purpose` in the code but just remembered that it is appended by calypso.

gikaragia commented on client/landing/stepper/declarative-flow/internals/steps-repository/sensei-purpose/index.tsx:
```diff
@@ -0,0 +1,158 @@
+
+type FormState = {
+	selected: string[];
+	other: string;
```
Maybe 'otherOption' or 'otherPurpose' is better?

yscik replied:
This is the schema for the Sensei setup wizard step: https://github.com/Automattic/sensei/blob/ab34051a8e7a79a5a5660d4df07364d8ee87568d/includes/admin/class-sensei-setup-wizard.php#L41-L44

I agree about the generic names in principle, but since the component is scoped to just selecting purposes I'd keep it like this to avoid having to map it to a different schema.

gikaragia replied:
I don't think that a schema that is defined in a different codebase should affect the implementation details here. I imagine that these two codebases will evolve separately from now on.

In comparison though, the values in the setup wizard are enclosed in a 'purpose' object which makes it easier to understand what the values refer to. Here we name the type FormState which is more generic.

yscik replied:
Updated the naming a bit in [3aa8fc4](https://github.com/Automattic/wp-calypso/pull/72582/commits/3aa8fc4b86ffa9904eef3fe4de4cb3e49fc5af4e), and also made the selected - wasSelected code a bit clearer.

gikaragia reviewed (commented):
Looks good! The only issue that I see is that the animation of the progress bar, is going smoothly and not in 'steps' as the rest of the progress bars that we have. 

But we can handle this in the other PR that you have open as I imagine it will modify this part.

gkaragia commented:
A couple of display issues:

![Screenshot 2023-02-02 at 15 06 50](https://user-images.githubusercontent.com/53191348/216333209-f7eae159-f3e4-4cc1-8b4a-a303c3374685.png)

- I find the border around the checkmark excessive
- There is an extra shadow on a selected option's border, to the right and bottom.
- I don't think that we need the border around the description.

yscik committed:
Adjust styling for purpose step

yscik committed:
Fix title alignment

yscik commented:
Adjusted the styles a bit:
- Change focus color from black and use focus-visible so it doesn't show on click
- Switch to 2px borders (1.5px produced the issue with certain resolutions I think)
- Subtitle also fixed to be black, it was a CSS issue

<img width="967" alt="image" src="https://user-images.githubusercontent.com/176949/216380418-70f825a1-a610-432d-b8e1-6500e5bec4f8.png">

Most of this is just a copy from Sensei's onboarding wizard, with some tuning to match the Wpcom styles - eg. Goals screen from a core step:
<img width="1115" alt="image" src="https://user-images.githubusercontent.com/176949/216381336-dc627761-0f4e-4af5-b904-d846fac6ad87.png">

gkaragia commented:
Tested the fixes and they work great!

yscik committed:
Update naming

gikaragia reviewed (approved):
Looks great!

gkaragia commented:
There were some failing E2E tests but they seemed unrelated. Restarted the job.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7799378

Hi @yscik, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Sell courses and generate income`
- `Provide certification`
- `Educate students`
- `Train employees`
- `Choose the purpose of your site`
- `Select your goals for offering courses, and we will help you set everything up.`
- `Order was completed successfully`
- `Preparing your new site`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7799378).
